sources/rottweiler/src/system.rs

```diff
@@ -0,0 +1,207 @@
+use serde::Deserialize;
+use snafu::prelude::*;
+use std::fs;
+use std::io::Write;
+use std::process::{Command, Stdio};
+
+type Result<T> = std::result::Result<T, snafu::Whatever>;
+
+const SYSTEMD_CREDS: &str = "/usr/bin/systemd-creds";
+const SYSTEMD_CRYPTSETUP: &str = "/usr/lib/systemd/systemd-cryptsetup";
+const CRYPTSETUP: &str = "/usr/sbin/cryptsetup";
+
+/// Encrypt data using systemd-creds with TPM2 PCRs
+pub fn systemd_creds_encrypt(name: &str, plaintext: &[u8]) -> Result<Vec<u8>> {
+    let pcrs = format!("--tpm2-pcrs={}", get_tpm2_pcrs()?);
+    execute(
+        SYSTEMD_CREDS,
+        &[
+            "encrypt",
+            "-",
+            "-",
+            "--name",
+            name,
+            "--with-key=tpm2",
+            &pcrs,
+        ],
+        Some(plaintext),
+    )
+}
+
+/// Decrypt data using systemd-creds with TPM2
+pub fn systemd_creds_decrypt(name: &str, ciphertext: &[u8]) -> Result<Vec<u8>> {
+    execute(
+        SYSTEMD_CREDS,
+        &["decrypt", "-", "-", "--name", name],
+        Some(ciphertext),
+    )
+}
+
+/// Attach a LUKS device using systemd-cryptsetup with the provided key
+pub fn systemd_cryptsetup_attach(
+    volume_name: &str,
+    source_device: &str,
+    key_data: &[u8],
+) -> Result<()> {
+    execute(
+        SYSTEMD_CRYPTSETUP,
+        &[
+            "attach",
+            volume_name,
+            source_device,
+            "/dev/stdin",
+            "luks,headless=yes,tries=1",
+        ],
+        Some(key_data),
+    )?;
+    Ok(())
+}
+
+/// Detach a LUKS device using systemd-cryptsetup
+pub fn systemd_cryptsetup_detach(volume_name: &str) -> Result<()> {
+    execute(SYSTEMD_CRYPTSETUP, &["detach", volume_name], None)?;
+    Ok(())
+}
+
+/// Format a device with LUKS2 using the provided key
+pub fn cryptsetup_luks_format(device: &str, key_data: &[u8]) -> Result<()> {
+    // Use minimal PBKDF2 iterations (1000, per NIST SP 800-132) since we're using a high-entropy
+    // TPM2-sealed key. This matches systemd's behavior and avoids unnecessary key stretching.
+    execute(
+        CRYPTSETUP,
+        &[
+            "luksFormat",
```

Comment:
I'd wondered if detecting the physical sector size of the underlying device is a good idea. Instance types with 4K physical sectors could perform better if we format to the same sector size.

On the contrary, though, if you re-use an EBS snapshot you could end up in a situation where the new host has a smaller physical sector size. My understanding is that having larger LUKS sectors than the physical disk sectors could lead to data corruption on power loss.